File: direct-file/src/main/scala/gov/irs/directfile/inputs/CollectionItemReference.scala
package gov.irs.directfile.inputs

import gov.irs.factgraph.compnodes.CollectionItemNode
import gov.irs.factgraph.FactDefinition
import gov.irs.formbuilder.exceptions.InvalidFormConfig
import gov.irs.formbuilder.parser.{ Input, InputContext, InputParser }

/** `<input type="collection-item-reference"/>` — which item of another collection this fact points at.
  *
  * Thirteen of Direct File's questions are this one shape: whose W-2 is this, whose 1099-G, which of the people in the
  * household qualifies you for head of household. The fact behind each is a `<Writable><CollectionItem
  * collection="…"/></Writable>`, and the answer is one collection item's id.
  *
  * ## Two things vary, and they come from different places
  *
  * **Which collection** is read out of the fact dictionary here rather than written in the Flow XML. The dictionary
  * already says it — `<CollectionItem collection="/filers"/>` becomes a `CollectionItemNode` whose alias is that path —
  * and a second copy in the flow would be a second thing to keep right. Resolving it at parse time also means a fact
  * that is not a collection reference fails the build with the path in the message, rather than rendering a question
  * with no options in it.
  *
  * **What to call each item** cannot come from the dictionary, because it is prose: upstream labels the options from
  * `fields.{path}.item`, a string of fact references evaluated once per item — `{{/filers/ * /firstName}} {{/filers/ *
  * /lastName}}`. That is authored content, so it travels in the flow as `item-label` on the `<input>` and the browser
  * substitutes each item's id into it (see `website-static/js/inputs/collection-item-reference.js`).
  *
  * Note that the label's collection need not be the referenced one. `/hohQualifyingPerson` references
  * `/hohQualifyingPeople`, which is a `<Filter>` over `/familyAndHousehold`, and its label names
  * `{{/familyAndHousehold/ * /firstName}}`. The ids are the same ids either way — a filter selects from its source
  * rather than re-keying it — so the substitution resolves. This is why the label is carried whole rather than as a
  * list of leaf names to hang off the collection path.
  *
  * Like [[Address]] and [[BankAccount]] this is a fieldset over several controls, so it renders its own label and
  * `suppliesOwnLabel` is true.
  */
object CollectionItemReference extends InputParser {

  val name = "collection-item-reference"

  override def parse(context: InputContext): Input = {
    val itemLabel = context.inputNode \@ "item-label"
    if (itemLabel.isEmpty) {
      throw InvalidFormConfig(s"<input type=\"$name\"/> needs an item-label attribute, at question ${context.path}")
    }

    Input.custom(
      name = name,
      optional = context.optional,
      templateVariables = Map("collection" -> collectionOf(context), "itemLabel" -> itemLabel),
      nodeType = Some("CollectionItemNode"),
      suppliesOwnLabel = true,
    )
  }

  /** The collection this question's fact refers into, from its `<Writable><CollectionItem collection="…"/>`.
    *
    * `fg-set` checks `nodeType` separately and would already have rejected a fact of the wrong type, but it does so
    * against the string form of the node. This reads the alias off the node itself, so the two cannot disagree — and an
    * alias-less `CollectionItemNode`, which the engine allows for a derived one, is caught here rather than reaching
    * the browser as a picker over nothing.
    */
  private def collectionOf(context: InputContext): String =
    context.factDictionary(context.path) match {
      case definition: FactDefinition =>
        definition.value match {
          case node: CollectionItemNode =>
            node.alias
              .map(_.toString)
              .getOrElse(
                throw InvalidFormConfig(
                  s"${context.path} is a collection item that names no collection, so " +
                    s"<input type=\"$name\"/> has nothing to list",
                ),
              )
          case other =>
            throw InvalidFormConfig(
              s"<input type=\"$name\"/> needs a <CollectionItem> fact; ${context.path} is a ${other.getClass.getSimpleName}",
            )
        }
      // `FactDictionary.apply` returns `FactDefinition | Null`, so this is the not-declared case and
      // nothing else — spelled `null` rather than `_` because the compiler can see that too.
      case null =>
        throw InvalidFormConfig(
          s"<input type=\"$name\"/> names ${context.path}, which the fact dictionary does not declare",
        )
    }
}

File: direct-file/src/main/resources/direct-file/website-static/js/inputs/collection-item-reference.js
// The browser half of `<input type="collection-item-reference"/>`: one radio per item of another
// collection, and the answer is the chosen item's id.
//
// The Scala half (`inputs/CollectionItemReference.scala`) resolves *which* collection out of the
// fact dictionary and puts it on the fieldset as `data-collection`; the flow carries the label
// pattern as `data-item-label`. Everything else is here, because how many options there are is a
// property of the graph in this tab rather than of the flow that was built.
//
// ## Why the options are cloned rather than rendered
//
// `<fg-collection>` already builds its rows this way, for the same reason: the count is not known
// until the page runs. The `<template class="df-item-reference__option">` in the Thymeleaf template
// is the one radio, so the markup still lives in HTML and this file wires it — the repo's rule
// (ADR-001), not an exception to it.
//
// ## Why an fg-update listener, when no other input type needs one
//
// The other custom types read one fact and write one control. This one's *options* are facts:
// `/hohQualifyingPerson` lists `/hohQualifyingPeople`, a `<Filter>` over the household, so
// answering a question elsewhere on the same page adds and removes options here.
// `setInputValueFromFactValue` runs on connect and not on fg-update, so nothing else would notice.
// The listener unsubscribes itself the first time it fires for an element that has left the
// document, which is what a collection row does when it is removed — registerInputType has no
// detach hook to do it from.

import { fg } from '../../vendor/form-builder/flow-runtime/js/fact-graph-engine.js'
import { factGraph } from '../../vendor/form-builder/flow-runtime/js/fg-fact-graph.js'
import { makeCollectionIdPath } from '../../vendor/form-builder/flow-runtime/js/fg-collection-utils.js'
import { registerInputType } from '../../vendor/form-builder/flow-runtime/js/input-types.js'

/** `{{/filers/*\/firstName}}` — one fact reference inside an authored label. */
const FACT_REFERENCE = /\{\{(\/[^}]+)\}\}/g

function group (el) {
  return el.querySelector('fieldset.df-item-reference')
}

function radios (el) {
  return group(el).querySelectorAll('input[type="radio"]')
}

/** The ids currently in the referenced collection. Empty rather than throwing on a path the graph
 *  cannot resolve yet — a collection that has no items is a picker with no options, not an error. */
function itemIds (fieldset) {
  try {
    return factGraph.getCollectionIds(fieldset.dataset.collection)
  } catch (error) {
    console.error(`Could not list ${fieldset.dataset.collection}:\n`, error)
    return []
  }
}

/**
 * One item's label: the authored pattern with each `{{/path/*\/leaf}}` replaced by that item's value.
 *
 * The `*` is substituted with this item's id, not with the id of any collection the question sits
 * inside — see CollectionItemReference.scala for why the pattern may name a different collection
 * than the one being listed.
 */
function labelFor (fieldset, id) {
  const text = fieldset.dataset.itemLabel.replace(FACT_REFERENCE, (_, path) => {
    try {
      const result = factGraph.get(makeCollectionIdPath(path, id))
      return result.hasValue ? result.get.toString() : ''
    } catch (error) {
      console.error(`Could not read ${path} for collection item ${id}:\n`, error)
      return ''
    }
  })
  // An item whose naming facts are all unanswered would otherwise be a radio with no words beside
  // it. Collapsing the whitespace first is what makes "{{firstName}} {{lastName}}" with neither
  // answered reach that fallback rather than arriving as a single space.
  return text.replace(/\s+/g, ' ').trim() || fieldset.dataset.unnamed
}

/** Rebuild the radios from the collection as it stands. Idempotent: the options are replaced. */
function build (el) {
  const fieldset = group(el)
  const container = fieldset.querySelector('.df-item-reference__options')
  const option = fieldset.querySelector('template.df-item-reference__option')
  const ids = itemIds(fieldset)

  container.replaceChildren()
  for (const id of ids) {
    const row = option.content.firstElementChild.cloneNode(true)
    const radio = row.querySelector('input')
    const label = row.querySelector('label')

    radio.id = `${fieldset.id}-${id}`
    // One name for the group, so the radios are mutually exclusive — and the fieldset's id rather
    // than the fact path, because inside a collection loop the id has had the enclosing item's id
    // spliced into it and the path attribute has too. Two rows of the same loop on one page would
    // otherwise share a name and answer each other.
    radio.name = fieldset.id
    radio.value = id
    radio.required = el.getAttribute('optional') !== 'true'
    label.htmlFor = radio.id
    label.textContent = labelFor(fieldset, id)

    container.appendChild(row)
  }

  fieldset.dataset.builtFor = ids.join(' ')
  return ids
}

/** The id the graph holds for this question, or '' — placeholders deliberately do not count. */
function answeredId (el) {
  const result = factGraph.get(el.getAttribute('path'))
  return result.complete ? (result.get?.idString ?? '') : ''
}

function check (el, id) {
  for (const radio of radios(el)) radio.checked = radio.value === id
}

registerInputType('collection-item-reference', {
  read (el) {
    const checked = group(el).querySelector('input[type="radio"]:checked')
    // '' is "unanswered", which is what makes fg-set delete the fact rather than write a blank.
    if (!checked) return ''
    // Built rather than handed over as a string, for the reason masked-number.js gives: the
    // string overload of factGraph.set knows six types and CollectionItem is not one of them.
    const built = fg.CollectionItemReferenceFactory(checked.value, group(el).dataset.collection, factGraph)
    return built.isRight ? built.right : ''
  },

  write (el, value, fact) {
    check(el, fact?.complete ? (fact.get?.idString ?? '') : '')
  },

  clear (el) {
    check(el, '')
  },

  attach (el) {
    build(el)

    group(el).addEventListener('change', (event) => {
      if (event.target.type !== 'radio') return
      el.clearValidationError()
      el.onChange()
    })

    // Re-listed when the collection behind the options changes, keeping whatever is answered.
    // Guarded on the id list so the common fg-update — a fact somewhere else on the page — costs a
    // string compare and no DOM.
    const onUpdate = () => {
      if (!el.isConnected) {
        document.removeEventListener('fg-update', onUpdate)
        return
      }
      const fieldset = group(el)
      if (itemIds(fieldset).join(' ') === fieldset.dataset.builtFor) return
      build(el)
      check(el, answeredId(el))
    }
    document.addEventListener('fg-update', onUpdate)
  }
})
